### PR TITLE
internal/usagestats: Stop sending url and arguments to bigquery

### PR DESCRIPTION
**Pull request by indradhanush**
Short term fix for issues encountered in #20628.

### CODE REVIEW

**Review by unknwon — APPROVED** [review]

File: internal/usagestats/event_handlers.go [hunk 1]
```diff
@@ -61,11 +61,9 @@ type bigQueryEvent struct {
 	AnonymousUserID string `json:"anonymous_user_id"`
 	FirstSourceURL  string `json:"first_source_url"`
 	UserID          int    `json:"user_id"`
-	URL             string `json:"url"`
 	Source          string `json:"source"`
 	Timestamp       string `json:"timestamp"`
 	Version         string `json:"version"`
-	Argument        string `json:"argument"`
```
Comment (rvantonder) [inline internal/usagestats/event_handlers.go:68]: My understanding of analytics (and I don't understand all of it) is that we keep types closely in sync with the big query schemas. For this event, the schema appears to be: https://github.com/sourcegraph/analytics/blob/master/BigQuery%20Schemas/dotcom_events.events.json. In other words, I think it might make sense to keep this type the way that it is, so that it is in sync with the schema, and use zero values to strip the data. Suggest @ebrodymoore can comment on this.

I am also wondering whether this will effectively stop other events (e.g., search latency events or any others really) from being recording in big query. I.e., whether our internals emit this events as this type, and the analytics scripts expect `Argument` to have certain values. E.g.,:

https://github.com/sourcegraph/analytics/blob/0a498958b349d7f6d0902e5f606e32fb96d8d676/BigQuery%20Schemas/search_latency_breakdown.rtf#L13

If so, we are effectively removing all analytics published through `argument`, which could be non-sensitive data like search latencies. In which case, I'm not sure this PR is good to merge. I have to read more code to do understand fully, but @ebrodymoore might have an immediate idea.

Comment (ebrodymoore) [reply internal/usagestats/event_handlers.go:68]: > In other words, I think it might make sense to keep this type the way that it is, so that it is in sync with the schema, and use zero values to strip the data.

This is a good callout. It's okay to have columns in the schema that are not sent in the payload, but it'd break BigQuery if something is sent in the payload that isn't in the schema. So in this case we're okay removing argument and URL. 

> If so, we are effectively removing all analytics published through argument, which could be non-sensitive data like search latencies.

You're absolutely correct about this. We are removing all analytics published through `argument`, including some with non-sensitive data, but this was actually my proposal in the first place. I had a Slack message drafted but never sent letting the search team (who is impacting the most by this) know about this change but obviously didn't communicate this change proactively to the relevant teams so that's on me. 

It does break a few charts, but we don't actually use most of them. This doesn't break anything involving pings so we'll still get the aggregate latency data on Cloud. 

What we could do is conditionally include `argument` so we still get search latency data, but again, I'm not aware of any active use from this so I think it's something we could add back in later. And for context on that, the private code launch is a really tight deadline and there are a lot of things going on, so I'm trying to make it as easy as possible for the core team.

**Comment by ebrodymoore** [comment]
Assuming Rijnard's comments are addressed, can we wait until a week or two before launch to do this? I'd really like to continue collecting argument and url up until when we have private code.

File: internal/usagestats/event_handlers.go (hunk 1 above)
Comment (rvantonder) [reply internal/usagestats/event_handlers.go:68]: Thanks for some context. Could you elaborate on these two sentences. These read contradictory but I'm not sure I follow so maybe it's not:

- [Removing] this doesn't break anything involving pings so we'll still get the aggregate latency data on Cloud.

vs.

- [ So  [that we ] we still get search latency data ] ... [ we could is conditionally include `argument` ] 

Will we get search latency data on Cloud or not? The first suggests "yes" the second seems to imply "no". I could probably read the code to understand what's going on here...

Comment (ebrodymoore) [reply internal/usagestats/event_handlers.go:68]: We'll still get aggregated latency (e.g. P99 for structural searches) but we won't be able to look at individual searches' latencies (e.g. the user searched ___ and it took 1.89 seconds), which we have the ability to look at now if we'd like. 

And we do publish latencies through `argument` as `latency: 1.89 seconds`

Comment (rvantonder) [reply internal/usagestats/event_handlers.go:68]: OK thanks that's clearer! I didn't/don't know of search latencies published the second way. Will try dig that up because I'm curious.

Comment (indradhanush) [reply internal/usagestats/event_handlers.go:68]: Thank you for your comments. It seems that even though removing `argument` breaks some metrics, but we are okay with it (for the time being) and removing it from the data type itself instead of sending empty values is a non issue as well? If so, may I resolve this conversation? Want to be sure I understood this right and am doing the right thing. :)

Comment (rvantonder) [reply internal/usagestats/event_handlers.go:68]: Yes, that sounds accurate, fine by me to resolve.

**Comment by tsenart** [comment]
@rvantonder, @indradhanush: There's one other problem with removing this these fields. We [rely on the data in the argument](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/search/backend.tsx?subtree=true#L722) field for the search home page data of "Recent X".

Screen Shot 2021-05-17 at 10 14 26

**Comment by rvantonder** [comment]
Uhhhh I think you're very right about that @tsenart. I happened to come across that usage last week but didn't make the connection to this PR. Oof. Difficult to say how to resolve this. Like, hypothetically we can create a different table/entries for this, but it seems like that means we're sidestepping the overall intent of RFC-358 because using this data can be construed as using/storing potentially private data (query): 

> We collect event/org/user-level data from individuals and organizations with private code on Cloud, but **never** any sensitive data. 

I will raise the issue in our search team sync happening in about 6.5 hours.

**Comment by indradhanush** [comment]
@tsenart Great catch. So if I understood this right, we can go ahead with the change of not publishing the data to big query but should still keep the field around in the `Event` data type? AFAIK, the link you shard above still reads from the PG database and not BigQuery? In that case we don't anticipate a side effect if we only stop pushing the data to BigQuery? :thinking:

**Comment by tsenart** [comment]
> we can go ahead with the change of not publishing the data to big query but should still keep the field around in the Event data type?

Yep, that makes sense!

**Comment by indradhanush** [comment]
@tsenart Upon taking a closer look at this, I realized we're changing the `bigQueryEvent` struct here. Which is [used only](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+bigQueryEvent&patternType=literal) in the `publishSourcegraphDotComEvent` function. 

The link you shared seems like is using the `Event` struct which is not being modified in this PR. Am I missing something? :thinking:

**Comment by tsenart** [comment]
> The link you shared seems like is using the Event struct which is not being modified in this PR. Am I missing something? 🤔

Nope, I'm the one who missed we're modifying big query specific code, not event logs themselves!

**Comment by rvantonder** [comment]
Correct--after reading @tsenart comment I also missed that this PR won't affect the internal logging, so good to go. Search team/ @lguychard to follow up separately on what our policy is for logging this data at all, to be in line with the philosophy of RFC 358.

**Comment by ebrodymoore** [comment]
Sorry for asking this now, and if this would take more than 10 minutes OR be a no-no for privacy/security then just tell me no. BUT could we conditionally allow the argument field to come through for a few specific events?

(and if the answer is this would be ok to do but it'd take more than 10 minutes, then cc @attfarhan 😄)

**Comment by indradhanush** [comment]
@ebrodymoore I'd say it would depend on the condition(s), but doing it in under 10 minutes would be a stretch nonetheless. 

Maybe you could share the conditions with us and @ryphil and @RafLeszczynski could coordinate on the priority + scope level for this iteration.

**Comment by ebrodymoore** [comment]
Whenever you all are ready to merge this then you should go ahead as @ryphil mentioned we're running up on the start of beta. Then we can think about adding something back.

Spec: if we can add a condition that says if event == 'SearchResultsFetched' then send argument, else don't send argument. These are nice-to-haves so I don't want to get in the way of the Cloud rollout, and will turn into really important data points for us to have back at some point. And of course there would have to be some controls to make sure these event arguments never contain sensitive information. The only two for what I would want right now is:
- SearchResultsFetched
- BrowserExtensionConnectedToServer

**Comment by indradhanush** [comment]
@ebrodymoore Got it. I've created an issue for this in our JIRA board and we will be taking it into account during our next planning session. :slightly_smiling_face:
